framework/src/media/utils/MediaUtils.cpp

```diff
@@ -43,6 +150,393 @@ static const std::string MP4_MIME_TYPE = "audio/mp4";
+		int ret = findTsPacket(demux->buffer + demux->offset, demux->buflen - demux->offset, false);
+		IF_FAIL_RETURN_VAL(ret >= UTILS_ERROR_NONE, ret);
+		demux->offset += ret;
+		uint8_t *tsPacket = demux->buffer + demux->offset;
+		demux->offset += TS_PACKET_SIZE;
+		if (TS_PID(tsPacket) != pid) {
+			continue; // PID not match
+		}
+		// unpack ts packet to get ES data
+		uint8_t  lenPayload = 0;
+		const uint8_t *ptrPayload = getPayloadData(tsPacket, &lenPayload);
+		if (!ptrPayload) {
+			continue; // no payload
+		}
+		if (TS_PAYLOAD_UNIT_START_INDICATOR(tsPacket)) {
+			// new PES packet start
+			if (PES_STREAM_ID(ptrPayload) < 0xc0 || PES_STREAM_ID(ptrPayload) > 0xdf) {
+				// not audio stream (stream id = 110xxxxx)
+				pes->continuity = 0xff;
+				continue;
+			}
+			pes->present = 0;
+			pes->packetLen = PES_PACKET_HEAD_BYTES + PES_PACKET_LENGTH(ptrPayload);
+			pes->esOffset = PES_PACKET_HEAD_BYTES + PES_STREAM_HEAD_BYTES + PES_HEADER_DATA_LENGTH(ptrPayload);
+		} else if (pes->continuity == 0xff) {
+			// waiting for new start indicator
+			continue;
+		} else if (TS_CONTINUITY_COUNTER(tsPacket) != ((pes->continuity + 1) % TS_CONTINUITY_COUNTER_MOD)) {
+			// continuity is broken
+			pes->continuity = 0xff;
+			total = 0; // drop copied ES data
+			continue;
+		}
+		// copy incoming ES data to output buffer
+		uint16_t header = (pes->present < pes->esOffset) ? pes->esOffset - pes->present : 0;
+		if (lenPayload > header) {
+			// ES data is available
+			uint16_t esLen = MINIMUM(pes->present + lenPayload, pes->packetLen) - pes->esOffset;
+			esLen = MINIMUM(size - total, esLen);
+			memcpy((void *)&buffer[total], (const void *)&ptrPayload[header], esLen);
+			total += esLen;
+			pes->esOffset += esLen;
+		} // else, no ES data in this ts packet
+		pes->continuity = TS_CONTINUITY_COUNTER(tsPacket);
+		pes->present += lenPayload;
+	}
+	return total;
+}
+
+/**
+ * @brief Parsing the mpeg2 transport stream data in buffer, to get audio stream codec type and other informations
+ * @param[in] buffer, buffer size, and audio type, channel, sample rate, pcm format adderss to receive.
+ * @return true - parsing success. false - parsing fail.
+ */
+static bool ts_parsing(unsigned char *buffer, unsigned int bufferSize, audio_type_t *audioType, unsigned int *channel, unsigned int *sampleRate, audio_format_type_t *pcmFormat)
+{
+	int ret;
+	ts_demux_t demux;
+
+	// init parser
+	ret = ts_demux_init(&demux, buffer, bufferSize);
+	if (ret < UTILS_ERROR_NONE) {
+		meddbg("Demux init failed, err %d\n", ret);
+		return false;
+	}
+
+	// parse and get PAT info
+	ret = ts_psi_process(&demux, PAT_PID);
+	if (ret < UTILS_ERROR_NONE) {
+		meddbg("PAT not found, err %d\n", ret);
+		return false;
+	}
+	psi_pat_info_t patInfo = {0};
+	ret = ts_psi_get_pat_info(&demux, &patInfo);
+	if (ret < UTILS_ERROR_NONE || patInfo.numOfPrgs < 1) {
+		meddbg("No program in PAT, err %d\n", ret);
+		return false;
+	}
+
+	// select the 1st program
+	uint16_t prgNum, pmtPid;
```

Comment:
how about initialize them, to avoid random value when ts_psi_get_program doesn't get value

Reply:
Fixed it, I check return value of ts_psi_get_program()